Optimization of DAG

div5252 opened the pull request:
1. Added optimize method for Real number, which user can call. Its implemented by traversal on shared pointers rather than reals, as traversal on real requires typecasting of shared pointers to real.
2. Added 2 variables - optimize_freq which denotes number of nodes after which every time we optimize. num_nodes maintain the number of nodes added in tree after last optimization. They are used in set_optimize_freq method. num_nodes is updated every time there is an operation like +=, -=, *=, /=.
3. Added method to find number of nodes, for it is required for finding num_nodes in real operation. 
4. Modified the print_tree method by traversing on shared pointers rather than reals, as traversal on real requires typecasting of shared pointers to real. Tested it and gave a huge improvement.
5. Corrected some typo's in check and distribute function. Also made the reals one and two in this function as static, so that they have the same address. Having same address will help us optimize such cases also.
6. Added two more variables - use_prev and prev_top which stores the result of previous optimization for next. This way we can simplify further two calls of optimization. 
7. Added tests for optimize() and set_optimize_freq() methods.

Laouen commented on include/real/real.hpp:
```diff
@@ -608,6 +650,256 @@ namespace boost {
+
+            // Checks if we have to optimize the Real
+            void check_if_optimize() { 
+                if (this->optimize_freq != -1 && this->num_nodes >= this->optimize_freq) {
```
because `num_nodes` is a counter from the last optimization I suggest using a more descriptive name even if it ends up longer. When I read `num_nodes` I spect having stored in the variable the total number of nodes within the tree, but this is not the case. A better name can be something like `num_nodes_without_optimization` or something describing the fact that the counter is reseated after each optimization.

div5252 replied:
How does `num_variables_after_last_optimize` sound?

Laouen replied:
Ok, that name sound mutch better. It''s kind of too long thought, but if there is no descriptive shorter name, then it is better to have a long descriptive name than a short confusing name.  There is that saying "written once and reader million times"

Laouen commented on test/real_optimization_test.cpp:
```diff
@@ -0,0 +1,343 @@
+            }
+            c.optimize();
+            CHECK(c == d);
+            CHECK(c.find_num_nodes() == 3);   // c is b + 5 * a
```
find_num_nodes shouldn't be 5? 3 leaf nodes (b, 5 and a) and 2 operation nodes (+ and *).

I think this question also answers my previous question about setting num_nodes = 2 in operations.

div5252 replied:
As I have answered above, it will have 3 variables.

Laouen commented on include/real/real.hpp:
```diff
@@ -608,6 +650,256 @@ namespace boost {
+                        positive = x.first;
+                        has_positive_node = true;
+                        first = false;
+                        break;
```
1. Why is this break here? It seems you are looking for the first positive count. Is this correct? Why? and if so, why not to use a std::find function for it? I understand that you are first looking the first positive count and then looping to process all counts? what is the role of the `first` boolean variable?

div5252 replied:
Let us take an example, 5*a+3*b-4*c. Now first the optimized tree would consist of multiplication of 5 and a, after that we also need to join 3*b by + and 4*c by -. If first expression is constructed, others need to be joined by +/-. Now it is also not necessary that first should be the first element of map, for in case if the count of element is 0, we don't add to our new tree.

Take another example, -a+b-c = (-1) * a + b + (-1) * c. Here number of nodes is 5. If we write it as b - a - c, number of nodes would be 3. In case it was all negative counts, i.e. -a-b-c, then there is no option but to simplify it as (-1)*a - b - c.

Yes, I can use std::find_if.

Laouen replied:
not sure to fully understand this. We will discuss it in the call this friday

Laouen reviewed (commented):
In general looks good, I added some minor comments and suggestions.

If you can, please write as a comment a simple explanation of the function `simplify` to be sure that I correctly understood the idea. This is the function you walk me through last week in the call right?

div5252 commented on include/real/real.hpp:
```diff
@@ -120,15 +139,21 @@ namespace boost {
                         std::shared_ptr<real_data<T>> rhs = std::make_shared<real_data<T>>(real_explicit<T>(denominator));
         
                         this->_real_p  = std::make_shared<real_data<T>>(real_operation(lhs, rhs, OPERATION::DIVISION));
+                        this->num_nodes = 2;
```
Here num_nodes refer to the number of variables in expression. Here it would be 2, as the parent node is an operation number.

Laouen replied:
Ok, perfect, but maybe the name num_nodes can be misunderstood as the number of nodes instead of the number of variables. Don't you think? maybe should be called num_vars?

div5252 commented on include/real/real.hpp:
```diff
@@ -90,7 +107,7 @@ namespace boost {
              * @param other - the boost::real::real instance to copy.
              */
-             real(const real<T>& other)  : _real_p(other._real_p) {};
+            real(const real<T>& other)  : _real_p(other._real_p), num_nodes(other.num_nodes) { this->check_if_optimize(); };
```
Do you think it is better if the optimization attributes are also transferred to the copies? If yes, then I will correct it.

Laouen replied:
I think yes. If you do a copy of an element, then, you want something exactly like that element, including his state. The optimization state is part or the element state.

div5252 commented on include/real/real.hpp:
```diff
@@ -71,7 +73,22 @@ namespace boost {
+                this->check_if_optimize();
+            };
+
+            // Frequency of optimizing the tree
```
@Laouen I don't think we can keep them in a new file or real_data because of this attribute - 
`std::pair<std::shared_ptr<real_data<T>>, std::map<std::shared_ptr<real_data<T>>, int>> prev_top;`
Maybe I will encapsulate all those 4 variables in a structure in real.hpp file.
What do you say?

div5252 replied:
Or I can make a structure in a new file `real_optimization_data.hpp` containing `real_data.hpp` file. And then include `real_optimization_data.hpp` in `real.hpp`.

Laouen replied:
Ok, definitively having a new file `real_optimizacion_data.hpp`  containing `real_data.hpp` is not a good idea from a semantical point of view because a real number is not in essence an optimized real number. But I think you can create the new file `real_optimization_data.hpp` with a structure containing the 4 optimization variables and import that file in this `real.hpp` file to use it in the same way as the real_data.hpp file is been used. Because there is no circular dependencies between the new `real_optimization_data.hpp` and the existing `real_data.hpp` you will be able to include `real_data.hpp` from the new `real_optimiation_data.hpp` file in order to declare the `std::pair<std::shared_ptr<real_data<T>>, std::map<std::shared_ptr<real_data<T>>, int>> prev_top;` attribute.

div5252 commented:
Changes made as requested - 
1. Made `optimization_data` structure in `real_optimization_data.hpp`.
2. Added better comments.
3. `simplify()` method improved by removing one unnecessary loop on the map elements.
4. Used `vars` instead of `nodes`, which is more appropriate term.
5. Added 2 helper methods - `increment_vars_and_check_if_optimize` and `set_vars_and_check_if_optimize`. 

@Laouen Now this PR is ready for merging, please review it.
